crates/leon/src/values.rs

```diff
@@ -0,0 +1,209 @@
+use std::{
+    borrow::Cow,
+    collections::{BTreeMap, HashMap},
+};
+
+pub trait Values {
+    fn get_value<'s, 'k: 's>(&'s self, key: &'k str) -> Option<Cow<'s, str>>;
+}
+
+impl<T> Values for &T
+where
+    T: Values,
+{
+    fn get_value<'s, 'k: 's>(&'s self, key: &'k str) -> Option<Cow<'s, str>> {
+        T::get_value(self, key)
+    }
+}
+
+impl Values for [(&str, &str)] {
+    fn get_value<'s, 'k: 's>(&'s self, key: &'k str) -> Option<Cow<'s, str>> {
+        self.iter().find_map(|(k, v)| {
+            if k == &key {
+                Some(Cow::Borrowed(*v))
+            } else {
+                None
+            }
+        })
+    }
+}
+
+impl<const N: usize> Values for [(&str, &str); N] {
+    fn get_value<'s, 'k: 's>(&'s self, key: &'k str) -> Option<Cow<'s, str>> {
+        self.iter().find_map(|(k, v)| {
+            if k == &key {
+                Some(Cow::Borrowed(*v))
+            } else {
+                None
+            }
+        })
+    }
+}
+
+impl Values for Vec<(&str, &str)> {
+    fn get_value<'s, 'k: 's>(&'s self, key: &'k str) -> Option<Cow<'s, str>> {
+        self.iter().find_map(|(k, v)| {
+            if k == &key {
+                Some(Cow::Borrowed(*v))
+            } else {
+                None
+            }
+        })
+    }
+}
+
+impl Values for HashMap<&str, &str> {
+    fn get_value<'s, 'k: 's>(&'s self, key: &'k str) -> Option<Cow<'s, str>> {
+        self.get(&key).map(|v| Cow::Borrowed(*v))
+    }
+}
+
+impl Values for BTreeMap<&str, &str> {
+    fn get_value<'s, 'k: 's>(&'s self, key: &'k str) -> Option<Cow<'s, str>> {
+        self.get(&key).map(|v| Cow::Borrowed(*v))
+    }
+}
+
+impl Values for [(String, &str)] {
```

Comment:
I think we should

```rust
impl<'a: 'it, 'it, It> Values for It
where
    It: Iterator<Item = (Cow<'a, str>, Cow<'a, str>)> + 'it,
```

And then also forward `impl Values for [(&str, &str)]` and this `impl Values for [(String, &str)]` to this iterator impl.

Reply:
I take it back since `Iterator::next` requires `&mut self` but here we only have immutable reference.

We could instead do this to avoid code duplication:

```rust
impl<K, V> Values for [(K, V)]
where
    K: AsRef<str>,
    V: AsRef<str>,
```